src/parser/helper/Helper.ts

```diff
@@ -12,50 +13,87 @@ export function formatCaptures(tree, captures) {
+    sourcePath: string,
+    excludedFolders: string[] = []
+): AsyncGenerator<ParseFile> {
+    try {
+        // Handle special case: if the specified sourcePath is a single file, just yield the file.
+        if ((await fs.promises.lstat(sourcePath)).isFile()) {
+            const parseFile = checkAndGetFileExtension(sourcePath);
+            if (parseFile !== undefined && fileExtensionToGrammar.has(parseFile.fileExtension)) {
+                yield parseFile;
+            }
+        } else {
+            // sourcePath points to a directory, so use recursive function to find all files.
+
+            // Create set to improve lookup performance:
+            const excludedSet = new Set(excludedFolders);
+            // The folder at sourcePath itself cannot be excluded, so continue using delegating yield* generator call:
+            yield* findFilesAsyncRecursive(sourcePath, excludedSet);
+        }
+    } catch (e) {
+        console.error(e);
     }
+}
 
-    fs.readdirSync(dir).forEach((file) => {
-        const fileExtension = file.split(".").pop();
+async function* findFilesAsyncRecursive(
+    dir: string,
+    excludedFolders: Set<string>
+): AsyncGenerator<ParseFile> {
+    try {
+        const openedDir = await fs.promises.opendir(dir);
 
-        const currentPath = path.join(dir, file);
-        if (fs.statSync(currentPath).isDirectory()) {
-            const isPathExcluded = excludedFolders.some((folder) => {
-                return path.basename(currentPath) === folder;
-            });
-            if (isPathExcluded) {
-                return;
-            }
-            fileList = findFilesRecursively(
-                currentPath,
-                supportedFileExtensions,
-                excludedFolders,
-                fileList
-            );
-        } else if (fileExtension && supportedFileExtensions.includes(fileExtension.toLowerCase())) {
-            const parseFile = getParseFile(currentPath);
-            if (parseFile != undefined) {
-                fileList = fileList.concat(parseFile);
-            }
-        }
-    });
+        for await (const currentEntry of openedDir) {
+            const currentPath = path.join(dir, currentEntry.name);
 
-    return fileList;
+            if (currentEntry.isDirectory()) {
+                const isPathExcluded = excludedFolders.has(currentEntry.name);
+                if (!isPathExcluded) {
+                    // The current directory is not excluded, so recurse into subdirectory,
+                    // using delegating yield* generator call:
+                    yield* findFilesAsyncRecursive(currentPath, excludedFolders);
+                }
+            } // End of if (isDirectory)
+            else {
+                const parseFile = checkAndGetFileExtension(currentPath);
+                if (
+                    parseFile !== undefined &&
+                    fileExtensionToGrammar.has(parseFile.fileExtension)
+                ) {
```

Comment:
```suggestion
                if (fileExtensionToGrammar.has(parseFile?.fileExtension)) {
```

Reply:
TypeScript dislikes this change, as "error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'."

Reply:
The set is probably defined to not allow undefined (that does make sense, it is accessing the entry, there is just none). We can keep the current part. I just like to write minimal code.